nixos/modules/services/web-apps/libretranslate.nix

```diff
@@ -0,0 +1,203 @@
+      };
+      port = lib.mkOption {
+        description = lib.mdDoc "LibreTranslate port.";
+        type = lib.types.port;
+        default = 5000;
+      };
+      dataDir = lib.mkOption {
+        description = lib.mdDoc "The data directory.";
+        type = lib.types.path;
+        default = "/var/lib/libretranslate";
+        example = "/opt/libretranslate";
+      };
+      threads = lib.mkOption {
+        description = lib.mdDoc "Set number of threads.";
+        type = lib.types.nullOr lib.types.int;
+        default = null;
+        example = 8;
+      };
+      enableApiKeys = lib.mkOption {
+        description = lib.mdDoc "Enable API keys database.";
+        type = lib.types.bool;
+        default = false;
+        example = true;
+      };
+      disableWebUI = lib.mkOption {
+        description = lib.mdDoc "Disable web ui.";
+        type = lib.types.bool;
+        default = false;
+        example = true;
+      };
+      updateModels = lib.mkOption {
+        description = lib.mdDoc "Update language models at startup";
+        type = lib.types.bool;
+        default = false;
+        example = true;
+      };
+      domain = lib.mkOption {
+        description = lib.mdDoc "The domain serving your LibreTranslate instance. Required for configure nginx as a reverse proxy.";
+        type = lib.types.str;
+        default = "";
+        example = "libretranslate.example.com";
+      };
+      configureNginx = lib.mkOption {
+        description = lib.mdDoc "Configure nginx as a reverse proxy for LibreTranslate.";
+        type = lib.types.bool;
+        default = false;
+      };
+      extraArgs = lib.mkOption {
+        description = lib.mdDoc "Extra arguments passed to the LibreTranslate.";
+        type = lib.types.listOf lib.types.str;
+        default = [ ];
+        example = [
+          "--debug"
+          "--disable-files-translation"
+        ];
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+
+    environment.systemPackages = [ ltmanageKeysCli ];
+
+    systemd.tmpfiles.rules = lib.mkIf (cfg.dataDir != "/var/lib/libretranslate") [
+      "d '${cfg.dataDir}' 0750 ${cfg.user} ${cfg.group} - -"
+      "z '${cfg.dataDir}' 0750 ${cfg.user} ${cfg.group} - -"
+    ];
+
+    systemd.services.libretranslate = {
+      description = "LibreTranslate service";
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+      environment = {
+        HOME = cfg.dataDir;
+      };
+      serviceConfig = lib.mkMerge [{
+        Type = "simple";
+        ExecStart = ''
+          ${cfg.package}/bin/libretranslate \
+            --host ${cfg.host} \
+            --port ${toString cfg.port} \
+            ${lib.optionalString cfg.enableApiKeys "--api-keys"} \
+            ${lib.optionalString cfg.disableWebUI "--disable-web-ui"} \
+            ${lib.optionalString cfg.updateModels "--update-models"} \
+            ${lib.optionalString (cfg.threads != null) "--threads ${toString cfg.threads}"} \
```

Comment:
Can we make extraArgs strucured and populate it with the arguments instead?

Reply:
Example?

Reply:
The setting type is https://github.com/NixOS/nixpkgs/blob/nixos-unstable/nixos/modules/services/matrix/synapse.nix#L192-L201 and then just concatStringsSep the option into the args string

Reply:
This variant gives me an error.
```diff
index 919469d5a19..99eceeaa548 100644
--- a/nixos/modules/services/web-apps/libretranslate.nix
+++ b/nixos/modules/services/web-apps/libretranslate.nix
@@ -2,6 +2,7 @@

 let
   cfg = config.services.libretranslate;
+  format = pkgs.formats.yaml {};
   ltmanageKeysCli = pkgs.writeShellScriptBin "ltmanage-keys" ''
     set -a
     export HOME="/var/lib/libretranslate"
@@ -85,13 +86,10 @@ in {
         description = lib.mdDoc "Configure nginx as a reverse proxy for LibreTranslate.";
       };
       extraArgs = lib.mkOption {
-        type = with lib.types; listOf str;
-        default = [ ];
-        example = [
-          "--debug"
-          "--disable-files-translation"
-        ];
-        description = lib.mdDoc "Extra arguments passed to the LibreTranslate.";
+        default = {};
+        type = lib.types.submodule {
+          freeformType = format.type;
+        };
       };
     };
   };
@@ -122,7 +120,7 @@ in {
             ${lib.optionalString cfg.disableWebUI "--disable-web-ui"} \
             ${lib.optionalString cfg.updateModels "--update-models"} \
             ${lib.optionalString (cfg.threads != null) "--threads ${toString cfg.threads}"} \
-            ${lib.concatStringsSep " " cfg.extraArgs}
+            ${lib.concatStringsSep " --" cfg.extraArgs}
         '';
         WorkingDirectory = cfg.dataDir;
         User = cfg.user;

```

```
building Nix...
building the system configuration...
error:
       … while calling the 'head' builtin

         at /home/user/nixpkgs/lib/attrsets.nix:820:11:

          819|         || pred here (elemAt values 1) (head values) then
          820|           head values
             |           ^
          821|         else

       … while evaluating the attribute 'value'

         at /home/user/nixpkgs/lib/modules.nix:768:9:

          767|     in warnDeprecation opt //
          768|       { value = builtins.addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          769|         inherit (res.defsFinal') highestPrio;

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: value is a set while a list was expected
```

Reply:
If this app uses GNU style arguments, maybe using [`lib.cli.toGNUCommandLineShell`](https://github.com/NixOS/nixpkgs/blob/master/lib/cli.nix) could be a viable solution? Type seems to be `with types; attrsOf str`

Reply:
@h7x4 that variant doesn't work either
```diff
diff --git a/nixos/modules/services/web-apps/libretranslate.nix b/nixos/modules/services/web-apps/libretranslate.nix
index 919469d5a193..12c31861b33d 100644
--- a/nixos/modules/services/web-apps/libretranslate.nix
+++ b/nixos/modules/services/web-apps/libretranslate.nix
@@ -85,7 +85,7 @@ in {
         description = lib.mdDoc "Configure nginx as a reverse proxy for LibreTranslate.";
       };
       extraArgs = lib.mkOption {
-        type = with lib.types; listOf str;
+        type = with lib.types; attrsOf str;
         default = [ ];
         example = [
           "--debug"
@@ -122,7 +122,7 @@ in {
             ${lib.optionalString cfg.disableWebUI "--disable-web-ui"} \
             ${lib.optionalString cfg.updateModels "--update-models"} \
             ${lib.optionalString (cfg.threads != null) "--threads ${toString cfg.threads}"} \
-            ${lib.concatStringsSep " " cfg.extraArgs}
+            ${lib.cli.toGNUCommandLineShell { } cfg.extraArgs}
         '';
         WorkingDirectory = cfg.dataDir;
         User = cfg.user;
```
Configuration:
```
    extraArgs = {
      "debug"
      "url-prefix translate"
    };
```
Error:
```
error:
       … while evaluating the attribute 'config'

         at /home/user/nixpkgs/lib/modules.nix:320:9:

          319|         options = checked options;
          320|         config = checked (removeAttrs config [ "_module" ]);
             |         ^
          321|         _module = checked (config._module);

       … while calling the 'seq' builtin

         at /home/user/nixpkgs/lib/modules.nix:320:18:

          319|         options = checked options;
          320|         config = checked (removeAttrs config [ "_module" ]);
             |                  ^
          321|         _module = checked (config._module);

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: syntax error, unexpected '"', expecting '.' or '='

       at /etc/nixos/configuration.nix:158:7:

          157|       "debug"
          158|       "url-prefix translate"
             |       ^
          159|     };
building Nix...
```

Reply:
I still think both of these solutions are viable, but the types in the examples you're providing are a bit faulty

In the first instance, you're calling `lib.concatStringsSep` which expects a list of strings, on `cfg.extraArgs` which is an attrset. I can't find the original example of how it was reduced down to a string, but I suspect the original type was a list of strings concatenated with spaces?

In the second, instance, it looks like you're trying to set `extraArgs` to something in-between of an attrset and a list. How about setting it like this?

```
    extraArgs = {
      debug = true;
      url-prefix = "translate";
    };
```

Reply:
Thank you very much, it works. I wouldn't have been able to come up with this code myself.